man/cupsaddsmb.man
.\"
.\" "$Id: cupsaddsmb.man 3041 2002-12-17 19:00:27Z swdev $"
.\"
.\"   cupsaddsmb man page for the Common UNIX Printing System (CUPS).
.\"
.\"   Copyright 1997-2003 by Easy Software Products.
.\"
.\"   These coded instructions, statements, and computer programs are the
.\"   property of Easy Software Products and are protected by Federal
.\"   copyright law.  Distribution and use rights are outlined in the file
.\"   "LICENSE.txt" which should have been included with this file.  If this
.\"   file is missing or damaged please contact Easy Software Products
.\"   at:
.\"
.\"       Attn: CUPS Licensing Information
.\"       Easy Software Products
.\"       44141 Airport View Drive, Suite 204
.\"       Hollywood, Maryland 20636-3111 USA
.\"
.\"       Voice: (301) 373-9603
.\"       EMail: cups-info@cups.org
.\"         WWW: http://www.cups.org
.\"
.TH cupsaddsmb 8 "Common UNIX Printing System" "27 September 2002" "Easy Software Products"
.SH NAME
cupsaddsmb \- export printers to samba for windows clients
.SH SYNOPSIS
.B cupsaddsmb
[ -H
.I samba-server
] [ -U
.I samba-user
] [ -h
.I cups-server
] [ -v ] -a
.br
.B cupsaddsmb
[ -H
.I samba-server
] [ -U
.I samba-user
] [ -h
.I cups-server
] [ -v ] printer [ ... printer ]
.SH DESCRIPTION
\fIcupsaddsmb\fR exports printers to the SAMBA software (version
2.2.0 or higher) for use with Windows clients. Depending on the
SAMBA configuration, you may need to provide a password to
export the printers. This program requires the CUPS or Adobe PostScript
printer driver files described below.
.LP
The \fI-H\fR option specifies the SAMBA server which defaults
to the CUPS server.
.LP
The \fI-U\fR option specifies the SAMBA print admin username which defaults
to your current username.
.LP
The \fI-a\fR option exports all known printers. Otherwise only
the named printers are exported.
.LP
The \fI-h\fR option specifies a different CUPS server to use.
.LP
The \fI-v\fR option specifies that verbose information should be
shown and is useful for debugging SAMBA configuration problems.
.SH SAMBA CONFIGURATION
\fIcupsaddsmb\fR uses the new RPC-based printing support in
SAMBA 2.2.x to provide printer drivers and PPD files to Windows
client machines. In order to use this functionality, you must
first configure SAMBA (via the smb.conf file) to support
printing through CUPS and provide a printer driver download
share, as follows:
.nf

    [global]
	load printers = yes
	printing = cups
	printcap name = cups

    [printers]
	comment = All Printers
	path = /var/spool/samba
	browseable = no
	public = yes
	guest ok = yes
	writable = no
	printable = yes
	printer admin = root

    [print$]
	comment = Printer Drivers
	path = /etc/samba/drivers
	browseable = yes
	guest ok = no
	read only = yes
	write list = root
.fi
.LP
This configuration assumes a FHS-compliant installation of
SAMBA; adjust the [printers] and [print$] share paths
accordingly on your system as needed.

.SH CUPS POSTSCRIPT DRIVERS FOR WINDOWS
\fIcupsaddsmb\fR can use the CUPS PostScript printer drivers for
Windows, which are available for download from the CUPS web
site.
.LP
The CUPS printer drivers are preferred over the Adobe drivers
since they support page accounting, banner pages, and page
labels on all printers.  However, currently only Windows NT,
2000, and XP are supported by the CUPS drivers, so you will also
need to get the Adobe driver to support Windows 95, 98, and Me
clients.
.LP
Once you have extracted the driver files, create a "drivers"
directory in the CUPS data directory (usually /usr/share/cups)
and copy the CUPS files using lowercase filenames, as follows:
.nf

    [Windows NT, 2000, and XP]
    cups.hlp
    cupsdrvr.dll
    cupsui.dll
.fi

.SH ADOBE POSTSCRIPT DRIVERS FOR WINDOWS
\fIcupsaddsmb\fR can use the Adobe PostScript printer drivers for
Windows, which are available for download from the Adobe web
site (http://www.adobe.com).
.LP
The CUPS printer drivers are preferred over the Adobe drivers
since they support page accounting, banner pages, and page
labels on all printers.
.LP
Once you have extracted the driver files, create a "drivers"
directory in the CUPS data directory (usually /usr/share/cups)
and copy the Adobe files using UPPERCASE filenames, as follows:
.nf

    [Windows 95, 98, and Me]
    ADFONTS.MFM
    ADOBEPS4.DRV
    ADOBEPS4.HLP
    DEFPRTR2.PPD
    ICONLIB.DLL
    PSMON.DLL

    [Windows NT, 2000, and XP]
    ADOBEPS5.DLL
    ADOBEPSU.DLL
    ADOBEPSU.HLP
.fi
.LP
You can download and copy either set of driver files
independently, and use the CUPS drivers for Windows NT, 2000,
and XP while using the Adobe drivers for the Windows 95, 98, and
Me clients.
.SH SEE ALSO
CUPS Software Administrators Manual,
http://localhost:631/documentation.html
.SH COPYRIGHT
Copyright 1993-2003 by Easy Software Products, All Rights Reserved.
.\"
.\" End of "$Id: cupsaddsmb.man 3041 2002-12-17 19:00:27Z swdev $".
.\"


systemv/cupsaddsmb.c
/*
 * "$Id$"
 *
 *   "cupsaddsmb" command for the Common UNIX Printing System (CUPS).
 *
 *   Copyright 2001 by Easy Software Products.
 *
 *   These coded instructions, statements, and computer programs are the
 *   property of Easy Software Products and are protected by Federal
 *   copyright law.  Distribution and use rights are outlined in the file
 *   "LICENSE.txt" which should have been included with this file.  If this
 *   file is missing or damaged please contact Easy Software Products
 *   at:
 *
 *       Attn: CUPS Licensing Information
 *       Easy Software Products
 *       44141 Airport View Drive, Suite 204
 *       Hollywood, Maryland 20636-3111 USA
 *
 *       Voice: (301) 373-9603
 *       EMail: cups-info@cups.org
 *         WWW: http://www.cups.org
 *
 * Contents:
 *
 *   main()             - Export printers on the command-line.
 *   do_samba_command() - Do a SAMBA command, asking for a password as needed.
 *   export_dest()      - Export a destination to SAMBA.
 */

/*
 * Include necessary headers...
 */

#include <stdio.h>
#include <stdlib.h>
#include <cups/cups.h>
#include <cups/string.h>
#include <errno.h>


/*
 * Local globals...
 */

int	Verbosity = 0;


/*
 * Local functions...
 */

int	do_samba_command(const char *, const char *, const char *);
int	export_dest(const char *);


/*
 * 'main()' - Export printers on the command-line.
 */

int			/* O - Exit status */
main(int  argc,		/* I - Number of command-line arguments */
     char *argv[])	/* I - Command-line arguments */
{
  int	i, j;		/* Looping vars */
  int	status;		/* Status from export_dest() */
  int	num_printers;	/* Number of printers */
  char	**printers;	/* Printers */


  for (i = 1; i < argc; i ++)
    if (strcmp(argv[i], "-a") == 0)
    {
      num_printers = cupsGetPrinters(&printers);

      for (j = 0, status = 0; j < num_printers; j ++)
        if ((status = export_dest(printers[j])) != 0)
	  break;

      for (j = 0; j < num_printers; j ++)
        free(printers[j]);

      if (num_printers)
        free(printers);

      if (status)
        return (status);
    }
    else if (strcmp(argv[i], "-U") == 0)
    {
      i ++;
      if (i >= argc)
      {
	puts("Usage: cupsaddsmb [-a] [-U user] [-v] [printer1 ... printerN]");
	return (1);
      }

      cupsSetUser(argv[i]);
    }
    else if (strcmp(argv[i], "-v") == 0)
      Verbosity = 1;
    else if (argv[i][0] != '-')
    {
      if ((status = export_dest(argv[i])) != 0)
	return (status);
    }
    else
    {
      puts("Usage: cupsaddsmb [-a] [-U user] [-v] [printer1 ... printerN]");
      return (1);
    }

  return (0);
}


/*
 * 'do_samba_command()' - Do a SAMBA command, asking for
 *                        a password as needed.
 */

int					/* O - Status of command */
do_samba_command(const char *command,	/* I - Command to run */
                 const char *args,	/* I - Argument(s) for command */
                 const char *filename)	/* I - File to use as input */
{
  int		status;			/* Status of command */
  char		temp[1024];		/* Command/prompt string */
  static const char *p = NULL;		/* Password data */


  for (status = 1;;)
  {
    if (p)
      snprintf(temp, sizeof(temp), "%s -N -U \'%s%%%s\' %s <%s",
               command, cupsUser(), p, args, filename ? filename : "/dev/null");
    else
      snprintf(temp, sizeof(temp), "%s -N -U \'%s\' %s <%s",
               command, cupsUser(), args, filename ? filename : "/dev/null");

    if (Verbosity)
    {
      printf("Running the following command:\n\n    %s\n", temp);

      if (filename)
      {
        char cat[1024];


        puts("\nwith the following input:\n");

        snprintf(cat, sizeof(cat), "cat %s", filename);
        system(cat);
      }
    }
    else
    {
      strncat(temp, " >/dev/null 2>/dev/null", sizeof(temp) - 1);
      temp[sizeof(temp) - 1] = '\0';
    }

    if ((status = system(temp)) != 0)
    {
      if (Verbosity)
        puts("");

      snprintf(temp, sizeof(temp),
               "Password for %s required to access %s via SAMBA: ",
	       cupsUser(), cupsServer());

      if ((p = cupsGetPassword(temp)) == NULL)
        break;
    }
    else
    {
      if (Verbosity)
        puts("");

      break;
    }
  }

  return (status);
}


/*
 * 'export_dest()' - Export a destination to SAMBA.
 */

int				/* O - 0 on success, non-zero on error */
export_dest(const char *dest)	/* I - Destination to export */
{
  int		status;		/* Status of smbclient/rpcclient commands */
  FILE		*fp;		/* File pointer for temp file */
  char		tempfile[1024];	/* Temporary file for print commands */
  const char	*ppdfile;	/* PPD file for printer drivers */
  char		command[1024];	/* Command to run */
  const char	*datadir;	/* CUPS_DATADIR */


  /* Get a temporary file for our smbclient and rpcclient commands... */
  cupsTempFile(tempfile, sizeof(tempfile));

  if ((datadir = getenv("CUPS_DATADIR")) == NULL)
    datadir = CUPS_DATADIR;

  /* Get the PPD file... */
  if ((ppdfile = cupsGetPPD(dest)) == NULL)
  {
    fprintf(stderr, "Warning: No PPD file for printer \"%s\"!\n", dest);
    return (1);
  }

  /* Write the smbclient commands needed for the Windows drivers... */
  if ((fp = fopen(tempfile, "w")) == NULL)
  {
    fprintf(stderr, "ERROR: Unable to create temporary file \"%s\" for export - %s\n",
	    tempfile, strerror(errno));
    unlink(ppdfile);
    return (2);
  }

  fputs("mkdir W32X86\n", fp);
  fprintf(fp, "put %s W32X86/%s.PPD\n", ppdfile, dest);
  fprintf(fp, "put %s/drivers/ADOBEPS5.DLL W32X86/ADOBEPS5.DLL\n",
          datadir);
  fprintf(fp, "put %s/drivers/ADOBEPSU.DLL W32X86/ADOBEPSU.DLL\n",
          datadir);
  fprintf(fp, "put %s/drivers/ADOBEPSU.HLP W32X86/ADOBEPSU.HLP\n",
          datadir);
  fputs("mkdir WIN40\n", fp);
  fprintf(fp, "put %s WIN40/%s.PPD\n", ppdfile, dest);
  fprintf(fp, "put %s/drivers/ADFONTS.MFM WIN40/ADFONTS.MFM\n",
          datadir);
  fprintf(fp, "put %s/drivers/ADOBEPS4.DRV WIN40/ADOBEPS4.DRV\n",
          datadir);
  fprintf(fp, "put %s/drivers/ADOBEPS4.HLP WIN40/ADOBEPS4.HLP\n",
          datadir);
  fprintf(fp, "put %s/drivers/DEFPRTR2.PPD WIN40/DEFPRTR2.PPD\n",
          datadir);
  fprintf(fp, "put %s/drivers/ICONLIB.DLL WIN40/ICONLIB.DLL\n",
          datadir);
  fprintf(fp, "put %s/drivers/PSMON.DLL WIN40/PSMON.DLL\n",
          datadir);
  fputs("quit\n", fp);

  fclose(fp);

  /* Run the smbclient command to copy the Windows drivers... */
  snprintf(command, sizeof(command), "smbclient //%s/print\\$", cupsServer());

  if ((status = do_samba_command(command, "", tempfile)) != 0)
  {
    fprintf(stderr, "ERROR: Unable to copy Windows printer driver files (%d)!\n",
            status);
    unlink(ppdfile);
    unlink(tempfile);
    return (3);
  }

  unlink(ppdfile);

  /* Write the rpcclient commands needed for the Windows drivers... */
  if ((fp = fopen(tempfile, "w")) == NULL)
  {
    fprintf(stderr, "ERROR: Unable to create temporary file \"%s\" for export - %s\n",
            tempfile, strerror(errno));
    unlink(tempfile);
    return (4);
  }

  fprintf(fp, "adddriver \"Windows NT x86\" "
              "\"%s:ADOBEPS5.DLL:%s.PPD:ADOBEPSU.DLL:ADOBEPSU.HLP:"
	      "NULL:RAW:NULL\"\n",
          dest, dest);
  fprintf(fp, "addprinter %s %s \"%s\" \"\"\n", dest, dest, dest);

 /*
  * MRS: For some reason, SAMBA doesn't like to install Win9x drivers
  *      with aux files.  They are currently commented out but further
  *      investigation is required...
  */

  fprintf(fp, "adddriver \"Windows 4.0\" "
              "\"%s:ADOBEPS4.DRV:%s.PPD:NULL:ADOBEPS4.HLP:"
	      "PSMON.DLL:RAW:NULL\"\n",
	      /*"PSMON.DLL:RAW:ADFONTS.MFM,DEFPRTR2.PPD,ICONLIB.DLL\"\n",*/
	      dest, dest);
  fputs("quit\n", fp);

  fclose(fp);

  /* Run the rpcclient command to install the Windows drivers... */
  if ((status = do_samba_command("rpcclient", cupsServer(), tempfile)) != 0)
  {
    fprintf(stderr, "ERROR: Unable to install Windows printer driver files (%d)!\n",
            status);
    unlink(tempfile);
    return (5);
  }

  unlink(tempfile);

  return (0);
}


/*
 * End of "$Id$".
 */
